src/ai/backend/common/exception.py

```diff
@@ -119,7 +122,89 @@ class VolumeUnmountFailed(RuntimeError):
     """
 
 
-class BackendError(web.HTTPError):
+class ErrorDomain(enum.StrEnum):
+    """
+    An enum to represent the domain of the error.
+    The domain is a string that represents the area where the error occurred.
+    """
+
+    BACKENDAI = "backendai"  # Whenever possible, use specific domain names instead of this one.
+    KERNEL = "kernel"
+    USER = "user"
+    SESSION = "session"
+    GROUP = "group"
+    IMAGE = "image"
+    INSTANCE = "instance"
+    ENDPOINT = "endpoint"
+    ROUTING = "routing"
+    DOTFILE = "dotfile"
+    VIRTUAL_FOLDER = "vfolder"
+
+
+class ErrorOperation(enum.StrEnum):
+    """
+    An enum to represent the operation where the error occurred.
+    The operation is a string that represents the action that was being performed
+    when the error occurred.
+    """
+
+    GENERIC = "generic"  # Whenever possible, use specific operation names instead of this one.
+    CREATE = "create"
+    READ = "read"
+    UPDATE = "update"
+    DELETE = "delete"
+    LIST = "list"
+    AUTH = "auth"
+    API = "api"
+    SERVICE = "service"
+
+
+class ErrorDetail(enum.StrEnum):
+    """
+    An enum to represent the specific error that occurred during the operation.
+    The error detail is a string that describes the specific error that occurred
+    during the operation.
+    """
+
+    INTERNAL_ERROR = (
+        "internal-error"  # Whenever possible, use specific error names instead of this one.
+    )
+    NOT_FOUND = "not-found"
+    ALREADY_EXISTS = "already-exists"
+    INVALID_PARAMETERS = "invalid-parameters"
+    TIMEOUT = "timeout"
+    FORBIDDEN = "forbidden"
+    UNAUTHORIZED = "unauthorized"
+    BAD_REQUEST = "bad-request"
+
+
+@dataclass
+class ErrorCode:
+    """
+    A dataclass to represent error codes for Backend.AI errors.
+
+    All fields are lowercase.
+
+    The top-level domain is set to backendai. Other domains represent the specific
+    area where the error occurred, such as kernel, user, session, etc.
+    Whenever possible, use detailed domain names instead of backendai of general purpose errors.
+
```

Comment:
```suggestion
    The top-level domain is set to `backendai`. Other domains represent the specific
    area where the error occurred, such as kernel, user, session, etc.
    Whenever possible, use detailed domain names instead of `backendai` of general purpose errors.

```